### PR TITLE
Implemented forward-backward function in probability.py

### PR DESCRIPTION
**Pull request by SnShine**
Changes:
* Added element_wise_product method to utils.py.
* While implementing forward_backward, I have used a new class HiddenMarkovModel. We can also use already implemented BayesNet class, but I think this implementation is pretty straight forward.
* Added tests for element_wise_product and forward_backward methods

And, @norvig, what have we decided regarding the function `update()`? I think using `self.a= a` is better than using `update(self, a=a)`.

### CODE REVIEW

File: utils.py [hunk 1]
```diff
@@ -162,11 +162,17 @@ def histogram(values, mode=0, bin_function=None):
         return sorted(bins.items())
 
 
-def dotproduct(X, Y):
+def sum_dotproduct(X, Y):
```
Comment (Schrodinger1926) [inline utils.py]: I think dotproduct should be fine

File: utils.py [hunk 2]
```diff
@@ -162,11 +162,17 @@ def histogram(values, mode=0, bin_function=None):
...
     """Return the sum of the element-wise product of vectors x and y."""
     return sum(x * y for x, y in zip(X, Y))
 
 
+def dotproduct(X, Y):
```
Comment (Schrodinger1926) [inline utils.py]: element wise multiplication and dotproduct are not the same thing.

**Comment by Schrodinger1926** [comment]
I too was writing forward-backward :) , Anyways nice work

File: probability.py [hunk 3]
```diff
@@ -524,11 +524,76 @@ def markov_blanket_sample(X, e, bn):
...
+    b = [1.0, 1.0]
+    bv = [b]    # we don't need bv; but we will have a list of all backward messages here
+    sv = [[0, 0] for i in range(len(ev))]
 
+    fv[0] = prior
+
+    for i in range(1, t+ 1):
+        fv[i] = forward(HMM, fv[i- 1], ev[i])
+    for i in range(t, -1, -1):
+        sv[i- 1] = normalize(dotproduct(fv[i], b))
```
Comment (Schrodinger1926) [inline probability.py]: isn't normalize a class method?

**Comment by SnShine** [comment]
Oops, my bad. What was I thinking?
@Schrodinger1926 Yeah, it's alright.

File: utils.py (hunk 2 above)
Comment (norvig) [inline utils.py]: `dotproduct` is a good name. See https://en.wikipedia.org/wiki/Dot_product

**Pull request merged**